internal/metrics/remote_write.go

```diff
@@ -0,0 +1,394 @@
+			}
+		} else {
+			log.Info("metrics range is empty")
+		}
+
+		// store last write
+		r.LastWrite = rangeEnd
+		err = ioutil.WriteFile(r.lastWriteFile, strconv.AppendInt(nil, r.LastWrite.UnixNano(), 10), 0600)
+		if err != nil {
+			log.Error(err)
+		}
+	}
+}
+
+// return value indicates whether or not to continue to next range
+// not necessarily that everything was written and successful
+func (r *RemoteWrite) writeRange(series []Series) bool {
+	requestSeries := make([]Series, 0, len(series)) // the series slice for current request
+	sampleIndex := 0                                // location within current series
+	numSamples := 0                                 // how many samples we accumulated for the request
+
+	for i := 0; i < len(series); {
+		currentSeries := series[i]
+		lenCurrentSeries := len(currentSeries.DataPoints)
+
+		if lenCurrentSeries != 0 {
+			if sampleIndex == 0 || len(requestSeries) == 0 { // first time we're using this series or requestSeries was reset
+				requestSeries = append(requestSeries, Series{Labels: currentSeries.Labels})
+			}
+
+			numMissingSamples := r.RequestNumSamples - numSamples
+			remainderOfSeries := lenCurrentSeries - sampleIndex
+			numSamplesToUse := numMissingSamples
+			if numMissingSamples > remainderOfSeries {
+				numSamplesToUse = remainderOfSeries
+			}
+
+			samplesToAppend := currentSeries.DataPoints[sampleIndex : sampleIndex+numSamplesToUse]
+			numSamples += numSamplesToUse
+			sampleIndex += numSamplesToUse
+
+			currReqSeries := &(requestSeries[len(requestSeries)-1])
+			currReqSeries.DataPoints = append(currReqSeries.DataPoints, samplesToAppend...)
+		}
+
+		if sampleIndex == lenCurrentSeries {
+			// advance to next series
+			i++
+			sampleIndex = 0
+		}
+
+		// if request not full and this is not the last series then continue to next series
+		if numSamples != r.RequestNumSamples && i < len(series) {
+			continue
+		}
+
+		if !r.writeRequest(requestSeries) {
+			return false
+		}
+
+		numSamples = 0
+		requestSeries = requestSeries[:0]
+	}
+
+	return true
+}
+
+// return value indicates whether or not to continue to next request
+// not necessarily that everything was written and successful
+func (r *RemoteWrite) writeRequest(series []Series) bool {
+	timeSeries, lowest, highest := toTimeSeries(series)
+
+	writeRequest := prompb.WriteRequest{
+		Timeseries: timeSeries,
+	}
+
+	if log.CurrentLevel() >= log.LevelTrace {
+		reqBytes, err := json.Marshal(writeRequest)
+		if err != nil {
+			log.Error("can not marshal prompb.WriteRequest to json", err)
+		}
+		log.Tracef("sending write request (lowest %s highest %s): %s", lowest.String(), highest.String(), string(reqBytes))
+	}
+
+	reqBytes, err := writeRequest.Marshal()
+	if err != nil {
+		log.Error("can not marshal prompb.WriteRequest", err)
+		return false
+	}
+
+	reqBytes = snappy.Encode(nil, reqBytes)
+
+	const numTries = 3
+	for try := 1; try <= numTries; try++ {
+		err = r.Client.Write(context.TODO(), reqBytes)
```

Comment:
Can we block here? What about timeout?

Reply:
Client is created with timeout. The timeout is part of RemoteWriteConfiguration. It will eventually come from control plane as part of remote server config.

Reply:
Here's the client code
![untitled (5)](https://user-images.githubusercontent.com/69590123/159285667-32ee8262-8315-4f79-8484-7130a0321860.png)

Reply:
How do we make sure that timeout * `RequestMaxSamples` is shorter than `waitInterval`?

Reply:
I am not sure what do u mean by timeout*RequestNumSamples. I will assume that u mean that 'time for all requests to server' is shorter than waitInterval.
The two are not related.
The waitInterval is a period of time to wait for sufficient data to be created in TSDB. This is AFTER we finish sending all the current data. So, one write iteration starts, it can take as much time as needed and then we wait 'waitInterval' for more data (notice that I do not use a ticker or timer but a 'sleep' between iterations).
I also use it as a period to wait for temporary errors to be resolved. This is because the interval is like infinity for the retries. Retry happens every X seconds and the interval in minutes.
I hope I was clear enough.

Reply:
@pkliczewski please review my answer